### PR TITLE
added new functions, fixed typos

### PR DESCRIPTION
**Pull request by sillyElvy**
## About

This PR adds new 2 new functions:
- Guild.modify_member()
- Member.modify()
- Member.add_to_thread()
- Channel.add_member

Also, it corrects some typos (inspired by #400 but corrected a bit more)
I also already included the correction of #400 so there should be no conflicts.

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [ ] I've made this pull request for/as: (check all that apply)
  - [x] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [x] Bugfix

### CODE REVIEW

File: interactions/api/http.py [hunk 1]
```diff
@@ -609,7 +609,7 @@ async def get_guild_widget_settings(self, guild_id: int) -> dict:
 
     async def get_guild_widget_image(self, guild_id: int, style: Optional[str] = None) -> str:
         """
-        Get a url representing a png image widget for the guild.
+        Get an url representing a png image widget for the guild.
```
Comment (Toricane) [inline interactions/api/http.py:612]: `a url` is correct

File: interactions/api/models/misc.py [hunk 2]
```diff
@@ -203,7 +203,7 @@ class Format:
     def stylize(cls, format: str, **kwargs) -> str:
         r"""
         This takes a format style from the object and
-        converts it into a useable string for ease.
+        converts it into an useable string for ease.
```
Comment (Toricane) [inline interactions/api/models/misc.py]: `a usable`
Useable *is* correct, but usable is used more, in both American and British English.

**Review by Toricane — CHANGES_REQUESTED** [review]
Here's some requested changes

File: interactions/api/http.py (hunk 1 above)
Comment (sillyElvy) [reply interactions/api/http.py:612]: Could you explain me why?

Comment (Toricane) [reply interactions/api/http.py:612]: It depends on the sound, not vowel, for whether you use an
URL is pronounced yoo-arr-ell, and the y sound makes it a url, not an url

https://www.latimes.com/socal/glendale-news-press/opinion/tn-gnp-a-word-please-theres-a-sound-reason-to-use-a-or-an-20150313-story.html 
here's an article I found that explains it

Comment (sillyElvy) [reply interactions/api/http.py:612]: I knew it depends on the sound but I didn't know that it's pronounced as "yoo"

Thanks!

File: interactions/api/models/channel.py [hunk 3]
```diff
@@ -317,6 +317,20 @@ async def modify(
...
+        member_id: int,
+    ) -> None:
+        """
+        This adds a member to the channel, if the channel is a thread
+
+        :param member_id: The id of the member to add to the channel
+        :type member_id: int
+        """
+        if not self.thread_metadata:
+            raise TypeError("The Channel you specified is not a thread!")
```
Comment (i0bs) [inline interactions/api/models/channel.py]:
```suggestion
            raise TypeError("The Channel you specified is not a thread!")  # TODO: Move to new error formatter.
```

Adds an internal note for the developers to flag for 2022 roadmap work.

File: interactions/api/models/guild.py [hunk 4]
```diff
@@ -655,6 +655,73 @@ async def modify_channel(
...
+        :type deaf: Optional[bool]
+        :param channel_id?: id of channel to move user to (if they are connected to voice). `None` to disconnect the user from its channel
+        :type channel_id: Optional[int]
+        :param communication_disabled_until?: when the user's timeout will expire and the user will be able to communicate in the guild again (up to 28 days in the future). `None` removes the timeout
+        :type communication_disabled_until: Optional[datetime.isoformat]
+        :param reason?: The reason of the modifying
+        :type reason: Optional[str]
+        """
+        member = Member(**await self._client.get_member(guild_id=int(self.id), member_id=member_id))
+        _nick = member.nick if nick == " " else nick
```
Comment (i0bs) [inline interactions/api/models/guild.py]:
```suggestion
        _nick = member.nick if nick is None else nick
```

Good python practice would be doing `nick=""` in the method when written which strictly tells for the nickname to be removed. `nick=None` is a bad practice.

File: interactions/api/models/guild.py [hunk 5]
```diff
@@ -655,6 +655,73 @@ async def modify_channel(
...
+        _roles = member.roles if not roles else roles
+        _communication_disabled_until = (
+            member.communication_disabled_until
+            if communication_disabled_until == 0
```
Comment (i0bs) [inline interactions/api/models/guild.py]:
```suggestion
            if communication_disabled_until is None
```

The same applies to how `nick` can be handled. Additionally, `communication_disabled_until` is an ISO8601 range and not an integer. Having a optional argument mutability as something not as `NoneType` or a sentinel-like value would be more problematic. We regulate `None` for this.

File: interactions/api/models/guild.py [hunk 6]
```diff
@@ -655,6 +655,73 @@ async def modify_channel(
         )
         return Channel(**res, _client=self._client)
 
+    async def modify_member(
+        self,
+        member_id: int,
+        nick: Optional[str] = " ",  # can't be None because None would remove the nickname
```
Comment (i0bs) [inline interactions/api/models/guild.py]:
```suggestion
        nick: Optional[str] = None,
```

File: interactions/api/models/guild.py [hunk 7]
```diff
@@ -655,6 +655,73 @@ async def modify_channel(
...
+        roles: Optional[List[int]] = None,
+        mute: Optional[bool] = None,
+        deaf: Optional[bool] = None,
+        channel_id: Optional[
+            int
+        ] = 0,  # can't be None because None would kick the member from the channel
```
Comment (i0bs) [inline interactions/api/models/guild.py]:
```suggestion
        ] = None
```

File: interactions/api/models/guild.py [hunk 8]
```diff
@@ -655,6 +655,73 @@ async def modify_channel(
...
+        communication_disabled_until: Optional[
+            datetime.isoformat
+        ] = 0,  # can't be None because None would remove the timeout
```
Comment (i0bs) [inline interactions/api/models/guild.py]:
```suggestion
        ] = None,
```

File: interactions/api/models/guild.py [hunk 9]
```diff
@@ -655,6 +655,73 @@ async def modify_channel(
...
+            else communication_disabled_until
+        )
+
+        payload = Member(
+            nick=_nick,
+            roles=_roles,
+            communication_disabled_until=_communication_disabled_until,
+        )
+
+        if channel_id != 0:
```
Comment (i0bs) [inline interactions/api/models/guild.py]:
```suggestion
        if channel_id:
```

This is a simpler approach to checking if a value is supplied from the argument to add to the payload.

File: interactions/api/models/member.py [hunk 10]
```diff
@@ -212,3 +212,84 @@ async def send(
 
         message = Message(**res, _client=self._client)
         return message
+
+    async def modify(
+        self,
+        guild_id: int,
+        nick: Optional[str] = " ",  # can't be None because None would remove the nickname
```
Comment (i0bs) [inline interactions/api/models/member.py]:
```suggestion
        nick: Optional[str] = None,
```

File: interactions/api/models/member.py [hunk 11]
```diff
@@ -212,3 +212,84 @@ async def send(
...
+        roles: Optional[List[int]] = None,
+        mute: Optional[bool] = None,
+        deaf: Optional[bool] = None,
+        channel_id: Optional[
+            int
+        ] = 0,  # can't be None because None would kick the member from the channel
```
Comment (i0bs) [inline interactions/api/models/member.py]:
```suggestion
        ] = None,
```

File: interactions/api/models/member.py [hunk 12]
```diff
@@ -212,3 +212,84 @@ async def send(
...
+        communication_disabled_until: Optional[
+            datetime.isoformat
+        ] = 0,  # can't be None because None would remove the timeout
```
Comment (i0bs) [inline interactions/api/models/member.py]:
```suggestion
        ] = None,
```

File: interactions/api/models/member.py [hunk 13]
```diff
@@ -212,3 +212,84 @@ async def send(
...
+        reason: Optional[str] = None,
+    ) -> "Member":
```
Comment (i0bs) [inline interactions/api/models/member.py:226]: You can additionally use `self` if you want, but this is fine for now.

File: interactions/api/models/member.py [hunk 14]
```diff
@@ -212,3 +212,84 @@ async def send(
...
+        :param deaf?: whether the user is deafened in voice channels
+        :type deaf: Optional[bool]
+        :param channel_id?: id of channel to move user to (if they are connected to voice). `None` to disconnect the user from its channel
+        :type channel_id: Optional[int]
+        :param communication_disabled_until?: when the user's timeout will expire and the user will be able to communicate in the guild again (up to 28 days in the future). `None` removes the timeout
+        :type communication_disabled_until: Optional[datetime.isoformat]
+        :param reason?: The reason of the modifying
+        :type reason: Optional[str]
+        """
+        _nick = self.nick if nick == " " else nick
```
Comment (i0bs) [inline interactions/api/models/member.py]:
```suggestion
        _nick = self.nick if nick is None else nick
```

File: interactions/api/models/member.py [hunk 15]
```diff
@@ -212,3 +212,84 @@ async def send(
...
+        _roles = self.roles if not roles else roles
+        _communication_disabled_until = (
+            self.communication_disabled_until
+            if communication_disabled_until == 0
```
Comment (i0bs) [inline interactions/api/models/member.py]:
```suggestion
            if communication_disabled_until is None
```

File: interactions/api/models/member.py [hunk 16]
```diff
@@ -212,3 +212,84 @@ async def send(
...
+            else communication_disabled_until
+        )
+
+        payload = Member(
+            nick=_nick,
+            roles=_roles,
+            communication_disabled_until=_communication_disabled_until,
+        )
+
+        if channel_id != 0:
```
Comment (i0bs) [inline interactions/api/models/member.py]:
```suggestion
        if channel_id:
```

File: interactions/api/models/member.pyi [hunk 17]
```diff
@@ -57,3 +57,23 @@ class Member(DictSerializerMixin):
         embeds=None,
         allowed_mentions=None,
     ) -> Message: ...
+
+    async def modify(
+        self,
+        guild_id: int,
+        nick: Optional[str] = " ",
```
Comment (i0bs) [inline interactions/api/models/member.pyi]:
```suggestion
        nick: Optional[str] = None,
```

File: interactions/api/models/member.pyi [hunk 18]
```diff
@@ -57,3 +57,23 @@ class Member(DictSerializerMixin):
...
+        roles: Optional[List[int]] = None,
+        mute: Optional[bool] = None,
+        deaf: Optional[bool] = None,
+        channel_id: Optional[
+            int
+        ] = 0,
```
Comment (i0bs) [inline interactions/api/models/member.pyi]:
```suggestion
        ] = None,
```

File: interactions/api/models/member.pyi [hunk 19]
```diff
@@ -57,3 +57,23 @@ class Member(DictSerializerMixin):
...
+        communication_disabled_until: Optional[
+            datetime.isoformat
+        ] = 0,
```
Comment (i0bs) [inline interactions/api/models/member.pyi]:
```suggestion
        ] = None,
```

File: interactions/api/models/message.py [hunk 20]
```diff
@@ -189,7 +189,7 @@ class Message(DictSerializerMixin):
     :ivar Optional[MessageInteraction] interaction?: Message interaction object, if the message is sent by an interaction.
     :ivar Optional[Channel] thread:? The thread that started from this message, if any, with a thread member object embedded.
     :ivar Optional[Union[Component, List[Component]]] components?: Components associated with this message, if any.
-    :ivar Optional[List[PartialSticker"]] sticker_items?: An array of message sticker item objects, if sent with them.
+    :ivar Optional[List["PartialSticker"]] sticker_items?: An array of message sticker item objects, if sent with them.
```
Comment (i0bs) [inline interactions/api/models/message.py]:
```suggestion
    :ivar Optional[List[PartialSticker]] sticker_items?: An array of message sticker item objects, if sent with them.
```

This was probably a mistake when I was doing the documentation. I took a lazy approach to copy-pasting the interfacing which introduced the quotations.

File: interactions/api/models/message.py [hunk 21]
```diff
@@ -189,7 +189,7 @@ class Message(DictSerializerMixin):
     :ivar Optional[MessageInteraction] interaction?: Message interaction object, if the message is sent by an interaction.
     :ivar Optional[Channel] thread:? The thread that started from this message, if any, with a thread member object embedded.
```
Comment (i0bs) [inline interactions/api/models/message.py]:
```suggestion
    :ivar Optional[Channel] thread?: The thread that started from this message, if any, with a thread member object embedded.
```

File: interactions/api/models/misc.py [hunk 22]
```diff
@@ -153,7 +153,7 @@ def epoch(self) -> float:
     @property
     def timestamp(self) -> datetime.datetime:
         """
-        The Datetime object variation of the the "Timestamp" field of the snowflake.
+        The Datetime object variation of the "Timestamp" field of the snowflake.
```
Comment (i0bs) [inline interactions/api/models/misc.py]:
```suggestion
        The Datetime object variation of the Timestamp field of the snowflake.
```

**Review by i0bs — CHANGES_REQUESTED** [review]
From what I've seen in other servers, there's a little bit of a misconception to how `None` checking optional arguments works. We always want to avoid using mutable default types where possible when it can stay complacent with how values are entered through conditionals. The none return has been used as a way to determine the equivalent of a "missing" argument.

We like to avoid doing `= " "`, `= 0` and alike where we're working directly to the API and isn't a client-facing aspect where we expect a core integral part to *need* catering. Our decorators somewhat break this but not really at the same time because we still use conditionals to check. This post goes into better detail on why we don't do this: https://nikos7am.com/posts/mutable-default-arguments/

Some of our code does stuff like this instead:

```py
# meth 1
def foo(bar: typing.Optional[str] = None):
    _bar = "" if bar is None else bar

# meth 2
def foo(bar: typing.Optional[str] = ""):
    _bar = bar if not bool(bar) else bar
```

For consistency reasons, we preferably want method 1.

**Comment by i0bs** [comment]
There's still a merge conflict

**Comment by sillyElvy** [comment]
> There's still a merge conflict

Fixed

**Review by i0bs — APPROVED** [review]

**Pull request merged**